### PR TITLE
Helm chart: Support latest api version of dashboard ingress

### LINKED ISSUE #2998: Usage of deprecated api for dashboard ingress
## Bug Report

**What version of Kubernetes are you using?**
1.21.9

**What version of Chaos Mesh are you using?**
2.1.3

**What did you do?**
I deployed chaos mesh via the helm chart. 

**What did you expect to see?**
I expected it to use the current version of ingresses

**What did you see instead?**
It deploys an ingress resource with API version extension/v1beta1 (see also https://github.com/chaos-mesh/chaos-mesh/blob/master/helm/chaos-mesh/templates/ingress.yaml#L16). 

**Why is this a problem**
We are using multiple ingress controllers. This is why we need to specify which controller handles the deployed ingress resource.  In the old version you could use an annotation to specify which ingress-class to use. The later version introduced a new field called 'ingressClassName'. By default ingress-nginx does not understand the old annotations anymore. Also we need to expect that the old version will be removed soon.

### PR DESCRIPTION
**Pull request by KimDoubleB**
Signed-off-by: Bobae Kim tree9295@gmail.com

### What problem does this PR solve?

Close #2998 

### What's changed and how it works?

- Added values to the `values.yaml` helm file for reference by `templates/ingress.yaml`.
  - `dashboard.ingress.apiVersionOverrides` field: apiVersion of ingress. This is used in `_helpers.tpl` to define apiVersion of ingress. 
  - `dashboard.ingress.ingressClassName` field: For define ingress controller 
  - `dashboard.ingress.paths` field: moved from `dashboard.ingress.hosts.paths`

- `chaos-dashboard.ingress.apiVersion` (`_helpers.tpl`) :  added helper so that ingress apiVersion is set according to k8s version.

### Related changes

- [ ] Need to update `chaos-mesh/website`
- [ ] Need to update `Dashboard UI`
- Need to **cheery-pick to release branches**
  - [ ] release-2.1
  - [ ] release-2.0

### Checklist

Tests

- [ ] Unit test
- [ ] E2E test
- [ ] No code
- [x] Manual test (add steps below)
  - I tested `helm template`, `helm install`

Side effects

- [ ] Breaking backward compatibility

### Release note 

```text
Please add a release note.

You can safely ignore this section if you don't think this PR needs a release note.
```

### DCO

If you find the DCO check fails, please run commands like below (Depends on the actual situations. For example, if the failed commit isn't the most recent) to fix it:

```shell
git commit --amend --signoff
git push --force
```

### CODE REVIEW

**Comment by fewdan** [comment]
Thank you for your contribution!

**Comment by KimDoubleB** [comment]
@fewdan 

Is there anything else I should do? 
I'm asking because it doesn't seem to be going on for a long time.

**Comment by fewdan** [comment]
> @fewdan Hi, Is there anything else I should do? I'm asking because it doesn't seem to be going on for a long time.

I will ask other reviewers to take a look.   

/cc @reviewer

**Review by STRRL — COMMENTED** [review]
Sorry for the late response

basically LGTM

and it seems we should update the changelog into the new `unreleased` section  :)

Thanks for your contribution!

**Comment by KimDoubleB** [comment]
@STRRL 
Thank you for your response!
Then I just have to wait without having to modify `CHANGELOG.md`, right?

**Comment by STRRL** [comment]
> @STRRL Thank you for your response! Then I just have to wait without having to modify `CHANGELOG.md`, right?

You should do something instead of waiting.

You should merge the latest changes from the master, and resolve the conflicts on `CHANGELOG.md`, then append the entry in `CHANGELOG.md`, under `Unreleased` section.

If I did not explain clearly, I think I could directly update the codes on your branch if you enabled the "Allow edits by maintainers"

**Comment by KimDoubleB** [comment]
> > @STRRL Thank you for your response! Then I just have to wait without having to modify `CHANGELOG.md`, right?
> 
> You should do something instead of waiting.
> 
> You should merge the latest changes from the master, and resolve the conflicts on `CHANGELOG.md`, then append the entry in `CHANGELOG.md`, under `Unreleased` section.
> 
> If I did not explain clearly, I think I could directly update the codes on your branch if you enabled the "Allow edits by maintainers"

Oh, I understand. So, I did rebase as you said and added change log to the unreleased section.

File: helm/chaos-mesh/templates/_helpers.tpl
Comment (MarkAndersonTrocme) [inline helm/chaos-mesh/templates/_helpers.tpl:168]: I had issues with this on eks, eks returns `1.21.9-eks-0d102a7` for the version and the semverCompare doesn't properly resolve.

Can you maybe use `Capabilities.KubeVersion.Major`?

Comment (MarkAndersonTrocme) [reply helm/chaos-mesh/templates/_helpers.tpl:168]: Or at least allow for a manual override with something like this: https://github.com/openfaas/faas-netes/blob/master/chart/openfaas/templates/_helpers.tpl#L22-L25

Comment (KimDoubleB) [reply helm/chaos-mesh/templates/_helpers.tpl]: @MarkAndersonTrocme 
Thank you for the review.
As you reviewed, I added `kubeVersionOverride` to the `values.yaml` and configured it to override the kubernetes version.

Comment (MarkAndersonTrocme) [reply helm/chaos-mesh/templates/_helpers.tpl]: Great, @STRRL just waiting on your review then 😄

Comment (STRRL) [reply helm/chaos-mesh/templates/_helpers.tpl]: Hi @KimDoubleB, I prefer to use `Capabilities.APIVersions.Has` with `networking.k8s.io/v1`, then fallback to `networking.k8s.io/v1beta1`, at the last fallback to `extensions/v1beta1"`

Comment (STRRL) [reply helm/chaos-mesh/templates/_helpers.tpl]: Hi @KimDoubleB, I prefer to use `Capabilities.APIVersions.Has` with `networking.k8s.io/v1`, then fallback to `networking.k8s.io/v1beta1`, at the last fallback to `extensions/v1beta1"`

Comment (STRRL) [reply helm/chaos-mesh/templates/_helpers.tpl]: rest LGTM ! PTAL @KimDoubleB 

cc @MarkAndersonTrocme

Comment (KimDoubleB) [reply helm/chaos-mesh/templates/_helpers.tpl]: @STRRL 
Thank you for the review. I modified the code. 
It's nice to see that it's more concise using `Capabilities.APIVersions.Has`.

Comment (KimDoubleB) [reply helm/chaos-mesh/templates/_helpers.tpl]: But, if you use `helm template` command, you must use option `--validate` (I used helm v3.9.0).

File: helm/chaos-mesh/templates/ingress.yaml
Comment (MarkAndersonTrocme) [inline helm/chaos-mesh/templates/ingress.yaml:81]: I don't quit understand the point of this else?

Comment (KimDoubleB) [reply helm/chaos-mesh/templates/ingress.yaml:81]: I didn't understand what you meant. To explain the above, if `ApiVersion` is not `networking.k8s.io/v1`, it is set to the ingress format of the previous version.

Comment (MarkAndersonTrocme) [reply helm/chaos-mesh/templates/ingress.yaml:81]: Something like this: https://gist.github.com/MarkAndersonTrocme/8d277794708249e10d5fa5bedd08aff3

Removing lines 53-73

Comment (MarkAndersonTrocme) [reply helm/chaos-mesh/templates/ingress.yaml:81]: @KimDoubleB what do you think?

Comment (STRRL) [reply helm/chaos-mesh/templates/ingress.yaml:81]: extract `eq (include "chaos-dashboard.ingress.apiVersion" $) "networking.k8s.io/v1"` into a vriable is much better not it is not required.

and it seems it is a little hard with the name thing to instead  `apiIsStable` 🙈

**Review by STRRL — APPROVED** [review]
LGTM

File: helm/chaos-mesh/templates/ingress.yaml
Comment (KimDoubleB) [reply helm/chaos-mesh/templates/ingress.yaml:81]: @MarkAndersonTrocme 

I think your example should have `host` field. Does it work without `host` field?

- ref: https://kubernetes.io/docs/concepts/services-networking/ingress/#ingress-rules

Comment (MarkAndersonTrocme) [reply helm/chaos-mesh/templates/ingress.yaml:81]: @KimDoubleB I have the host field in my example on [line 44](https://gist.github.com/MarkAndersonTrocme/8d277794708249e10d5fa5bedd08aff3#L44)

Comment (MarkAndersonTrocme) [reply helm/chaos-mesh/templates/ingress.yaml:81]: But feel free to merge it's just a nitpick

**Comment by MarkAndersonTrocme** [comment]
@STRRL Can this get merged please?

**Comment by STRRL** [comment]
> @STRRL Can this get merged please?

SORRY for the late response.  This PR still requires one more approvement.... I would ping more reviewers.

**Comment by STRRL** [comment]
/cc @iguoyr @Hexilee 

PTAL

**Comment by FuyuanChu** [comment]
Hello, any updates on this?

**Review by Hexilee — APPROVED** [review]
LGTM!

**Pull request merged**
